File: python/src/n0973_k_closest_points_to_origin.py
from heapq import heappush, heappushpop


def _distance(point: list[int]) -> int:
    return point[0] ** 2 + point[1] ** 2


# O(n log n)
def k_closest(points: list[list[int]], k: int) -> list[list[int]]:
    points.sort(key=_distance)
    return points[:k]


# O(n log k)
def k_closest_heap(points: list[list[int]], k: int) -> list[list[int]]:
    heap: list[tuple[int, list[int]]] = []
    for point in points:
        item = (-_distance(point), point)
        if len(heap) < k:
            heappush(heap, item)
        else:
            heappushpop(heap, item)
    return [point for [_, point] in heap]

File: typescript/src/n0973_k_closest_points_to_origin.ts
import { PriorityQueue } from '@datastructures-js/priority-queue';

const distance = ([x, y]: [number, number]) => x * x + y * y;

// O(n log n)
export function kClosest(
  points: [number, number][],
  k: number
): [number, number][] {
  points.sort((a, b) => distance(a) - distance(b));
  return points.slice(0, k);
}

// O(n log k)
export function kClosestHeap(
  points: [number, number][],
  k: number
): [number, number][] {
  const heap = new PriorityQueue<[number, number]>((a, b) => {
    return distance(b) - distance(a);
  });
  for (const point of points) {
    heap.push(point);
    if (heap.size() > k) {
      heap.pop();
    }
  }
  return heap.toArray();
}
